File: src/git.c
#define _POSIX_C_SOURCE 200809L
#include "git.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>
#include <sys/stat.h>

/* Helper to run git command and get output */
static char *run_git_command(const char *repo_path, const char *cmd) {
    /* Validate repo_path doesn't contain shell metacharacters that could escape quotes */
    if (repo_path && strchr(repo_path, '\'')) {
        return NULL;  /* Reject paths with single quotes to prevent command injection */
    }

    char full_cmd[1024];
    snprintf(full_cmd, sizeof(full_cmd), "cd '%s' && git %s 2>/dev/null", repo_path, cmd);

    FILE *fp = popen(full_cmd, "r");
    if (!fp) return NULL;

    char *result = NULL;
    size_t size = 0;
    size_t capacity = 1024;
    result = malloc(capacity);
    if (!result) {
        pclose(fp);
        return NULL;
    }

    char buffer[256];
    while (fgets(buffer, sizeof(buffer), fp)) {
        size_t len = strlen(buffer);
        if (size + len >= capacity) {
            capacity *= 2;
            char *new_result = realloc(result, capacity);
            if (!new_result) {
                free(result);
                pclose(fp);
                return NULL;
            }
            result = new_result;
        }
        memcpy(result + size, buffer, len);
        size += len;
    }

    pclose(fp);
    result[size] = '\0';
    return result;
}

char *git_find_root(const char *path) {
    if (!path) return NULL;

    char *test_path = strdup(path);
    if (!test_path) return NULL;

    /* If path is a file, get its directory */
    struct stat st;
    if (stat(test_path, &st) == 0 && !S_ISDIR(st.st_mode)) {
        /* dirname may modify buffer or return static storage - make a copy first */
        char *path_copy = strdup(test_path);
        if (!path_copy) {
            free(test_path);
            return NULL;
        }
        char *dir = dirname(path_copy);
        char *dir_copy = strdup(dir);
        free(path_copy);
        free(test_path);
        test_path = dir_copy;
        if (!test_path) return NULL;
    }

    /* Walk up directory tree looking for .git */
    while (test_path && strlen(test_path) > 1) {
        char git_dir[1024];
        snprintf(git_dir, sizeof(git_dir), "%s/.git", test_path);

        if (stat(git_dir, &st) == 0) {
            return test_path;  /* Found .git directory */
        }

        /* Go up one level */
        char *parent = dirname(test_path);
        if (strcmp(parent, test_path) == 0) {
            break;  /* Reached root */
        }

        char *new_path = strdup(parent);
        free(test_path);
        test_path = new_path;
    }

    free(test_path);
    return NULL;
}

bool git_is_repo(const char *path) {
    char *root = git_find_root(path);
    if (root) {
        free(root);
        return true;
    }
    return false;
}

char *git_get_branch(const char *repo_path) {
    char *output = run_git_command(repo_path, "branch --show-current");
    if (!output) return NULL;

    /* Remove trailing newline */
    size_t len = strlen(output);
    if (len > 0 && output[len - 1] == '\n') {
        output[len - 1] = '\0';
    }

    return output;
}

GitRepo *git_repo_open(const char *path) {
    if (!path) return NULL;

    char *root = git_find_root(path);
    if (!root) return NULL;

    GitRepo *repo = malloc(sizeof(GitRepo));
    if (!repo) {
        free(root);
        return NULL;
    }

    repo->root_path = root;
    repo->current_branch = git_get_branch(root);
    repo->is_repo = true;
    repo->file_statuses = NULL;
    repo->file_count = 0;

    return repo;
}

void git_repo_close(GitRepo *repo) {
    if (!repo) return;

    free(repo->root_path);
    free(repo->current_branch);

    for (size_t i = 0; i < repo->file_count; i++) {
        free(repo->file_statuses[i].filename);
    }
    free(repo->file_statuses);

    free(repo);
}

GitFileStatus *git_get_status(const char *repo_path, size_t *count) {
    if (!repo_path || !count) return NULL;

    char *output = run_git_command(repo_path, "status --porcelain");
    if (!output) {
        *count = 0;
        return NULL;
    }

    /* Count lines */
    size_t num_files = 0;
    for (char *p = output; *p; p++) {
        if (*p == '\n') num_files++;
    }

    if (num_files == 0) {
        free(output);
        *count = 0;
        return NULL;
    }

    GitFileStatus *statuses = malloc(sizeof(GitFileStatus) * num_files);
    if (!statuses) {
        free(output);
        *count = 0;
        return NULL;
    }

    /* Parse lines */
    size_t idx = 0;
    char *line = strtok(output, "\n");
    while (line && idx < num_files) {
        if (strlen(line) > 3) {
            statuses[idx].status = line[1] != ' ' ? line[1] : line[0];
            statuses[idx].filename = strdup(line + 3);
            idx++;
        }
        line = strtok(NULL, "\n");
    }

    free(output);
    *count = idx;
    return statuses;
}

GitDiff *git_get_file_diff(const char *repo_path, const char *filename) {
    if (!repo_path || !filename) return NULL;

    /* Get the file content line count first */
    char path[1024];
    snprintf(path, sizeof(path), "%s/%s", repo_path, filename);

    FILE *f = fopen(path, "r");
    if (!f) return NULL;

    size_t num_lines = 0;
    char buffer[4096];
    while (fgets(buffer, sizeof(buffer), f)) {
        num_lines++;
    }
    fclose(f);

    GitDiff *diff = malloc(sizeof(GitDiff));
    if (!diff) return NULL;

    diff->num_lines = num_lines;
    diff->line_statuses = calloc(num_lines, sizeof(GitLineStatus));
    if (!diff->line_statuses) {
        free(diff);
        return NULL;
    }

    /* Get git diff */
    char cmd[1024];
    snprintf(cmd, sizeof(cmd), "diff HEAD -- '%s'", filename);
    char *output = run_git_command(repo_path, cmd);

    if (!output) {
        /* No diff means file is unchanged or new */
        /* Check if file is tracked */
        snprintf(cmd, sizeof(cmd), "ls-files '%s'", filename);
        char *tracked = run_git_command(repo_path, cmd);
        if (tracked && strlen(tracked) > 0) {
            /* File is tracked but unchanged */
            free(tracked);
        } else {
            /* File is new (untracked or added) */
            for (size_t i = 0; i < num_lines; i++) {
                diff->line_statuses[i] = GIT_ADDED;
            }
        }
        free(tracked);
        return diff;
    }

    /* Parse diff output
     * Format: @@ -old_start,old_count +new_start,new_count @@
     */
    char *line = strtok(output, "\n");
    int current_line = 0;
    bool in_hunk = false;

    while (line) {
        if (strncmp(line, "@@", 2) == 0) {
            /* Parse hunk header */
            int new_start, new_count;
            if (sscanf(line, "@@ -%*d,%*d +%d,%d @@", &new_start, &new_count) == 2) {
                current_line = new_start - 1;  /* Convert to 0-based */
                in_hunk = true;
            }
        } else if (in_hunk && strlen(line) > 0) {
            char marker = line[0];
            if (marker == '+') {
                /* Added line - exists in new file */
                if (current_line < (int)num_lines) {
                    diff->line_statuses[current_line] = GIT_ADDED;
                    current_line++;
                }
            } else if (marker == '-') {
                /* Deleted line - doesn't exist in new file, don't mark or advance */
                /* These lines are in the old version but not in the new version */
            } else if (marker == ' ') {
                /* Unchanged line - advance position in new file */
                current_line++;
            }
        }
        line = strtok(NULL, "\n");
    }

    free(output);
    return diff;
}

void git_diff_free(GitDiff *diff) {
    if (!diff) return;
    free(diff->line_statuses);
    free(diff);
}

File: plugins/features/git.lua
-- Git Integration Plugin for OCCE
-- Pure Lua implementation using process.execute() API
-- This demonstrates moving hard-coded C features to plugins!

local M = {}

-- Cache for git information (per buffer)
local git_cache = {}

-- Helper: Execute git command in a directory
local function run_git_command(repo_path, cmd)
    if not repo_path then
        return nil
    end

    -- Security: Reject paths with single quotes
    if repo_path:find("'") then
        return nil
    end

    local full_cmd = string.format("cd '%s' && git %s 2>/dev/null", repo_path, cmd)
    local stdout, stderr, exitcode = process.execute(full_cmd)

    if exitcode == 0 and stdout then
        return stdout
    end

    return nil
end

-- Find git repository root by walking up directory tree
function M.find_root(path)
    if not path or path == "" then
        return nil
    end

    -- Check if this is a file - extract directory
    local is_file_cmd = string.format("test -f '%s' && echo 'file' || echo 'dir'", path)
    local result = process.execute(is_file_cmd)
    if result and result:match("file") then
        -- Get directory name
        local dir_cmd = string.format("dirname '%s'", path)
        local dir = process.execute(dir_cmd)
        if dir then
            path = dir:gsub("\n$", "")
        end
    end

    -- Walk up directory tree looking for .git
    local test_path = path
    while test_path and test_path ~= "/" and test_path ~= "" do
        local git_dir = test_path .. "/.git"
        local check_cmd = string.format("test -e '%s' && echo 'found' || echo 'not found'", git_dir)
        local found = process.execute(check_cmd)

        if found and found:match("found") then
            return test_path
        end

        -- Go up one level
        local parent_cmd = string.format("dirname '%s'", test_path)
        local parent = process.execute(parent_cmd)
        if parent then
            parent = parent:gsub("\n$", "")
            if parent == test_path then
                break  -- Reached root
            end
            test_path = parent
        else
            break
        end
    end

    return nil
end

-- Check if path is in a git repository
function M.is_repo(path)
    local root = M.find_root(path)
    return root ~= nil
end

-- Get current branch name
function M.get_branch(repo_path)
    local output = run_git_command(repo_path, "branch --show-current")
    if output then
        -- Remove trailing newline
        return output:gsub("\n$", "")
    end
    return nil
end

-- Get file status (M=modified, A=added, D=deleted, etc.)
function M.get_status(repo_path)
    local output = run_git_command(repo_path, "status --porcelain")
    if not output or output == "" then
        return {}
    end

    local statuses = {}
    for line in output:gmatch("[^\n]+") do
        if #line > 3 then
            local status = line:sub(2, 2)
            if status == " " then
                status = line:sub(1, 1)
            end
            local filename = line:sub(4)
            statuses[filename] = status
        end
    end

    return statuses
end

-- Get per-line diff for a file
function M.get_file_diff(repo_path, filename)
    if not repo_path or not filename then
        return nil
    end

    -- Get file line count
    local file_path = repo_path .. "/" .. filename
    local count_cmd = string.format("wc -l < '%s' 2>/dev/null || echo 0", file_path)
    local count_str = process.execute(count_cmd)
    local num_lines = tonumber(count_str) or 0

    if num_lines == 0 then
        return nil
    end

    -- Initialize all lines as unchanged
    local line_statuses = {}
    for i = 1, num_lines do
        line_statuses[i] = "unchanged"
    end

    -- Get git diff
    local diff_output = run_git_command(repo_path, string.format("diff HEAD -- '%s'", filename))

    if not diff_output or diff_output == "" then
        -- No diff - check if file is tracked
        local tracked = run_git_command(repo_path, string.format("ls-files '%s'", filename))
        if not tracked or tracked == "" then
            -- File is new (untracked)
            for i = 1, num_lines do
                line_statuses[i] = "added"
            end
        end
        -- else file is tracked but unchanged
        return line_statuses
    end

    -- Parse diff output
    -- Format: @@ -old_start,old_count +new_start,new_count @@
    local current_line = 0
    local in_hunk = false

    for line in diff_output:gmatch("[^\n]+") do
        if line:match("^@@") then
            -- Parse hunk header
            local new_start, new_count = line:match("@@ %-(%d+),(%d+) %+(%d+),(%d+) @@")
            if new_start then
                current_line = tonumber(new_start) - 1  -- Convert to 0-based
                in_hunk = true
            else
                -- Try simpler format: @@ -start +start @@
                new_start = line:match("@@ [^ ]+ %+(%d+) @@")
                if new_start then
                    current_line = tonumber(new_start) - 1
                    in_hunk = true
                end
            end
        elseif in_hunk and #line > 0 then
            local marker = line:sub(1, 1)
            if marker == "+" then
                -- Added line
                current_line = current_line + 1
                if current_line <= num_lines then
                    line_statuses[current_line] = "added"
                end
            elseif marker == "-" then
                -- Deleted line (doesn't exist in new file)
                -- Don't advance current_line
            elseif marker == " " then
                -- Unchanged line
                current_line = current_line + 1
            end
        end
    end

    return line_statuses
end

-- Initialize git for a buffer
function M.init_buffer(filename)
    if not filename then
        return nil
    end

    local root = M.find_root(filename)
    if not root then
        git_cache[filename] = nil
        return nil
    end

    -- Get relative path from repo root
    local rel_path = filename
    if filename:sub(1, #root) == root then
        rel_path = filename:sub(#root + 2)  -- +2 to skip the /
    end

    local cache = {
        root = root,
        branch = M.get_branch(root),
        rel_path = rel_path,
        diff = M.get_file_diff(root, rel_path)
    }

    git_cache[filename] = cache
    return cache
end

-- Get cached git info for buffer
function M.get_cache(filename)
    if not filename then
        return nil
    end

    local cache = git_cache[filename]
    if not cache then
        -- Try to initialize
        cache = M.init_buffer(filename)
    end

    return cache
end

-- Refresh git diff for current buffer
function M.refresh()
    local filename = buffer.get_filename()
    if filename then
        git_cache[filename] = nil
        return M.init_buffer(filename)
    end
    return nil
end

-- Gutter renderer function (called for each line)
function M.render_gutter(line_num)
    local filename = buffer.get_filename()
    if not filename then
        return "  "
    end

    local cache = M.get_cache(filename)
    if not cache or not cache.diff then
        return "  "
    end

    -- line_num is 0-based in C, convert to 1-based for Lua
    local lua_line = line_num + 1
    local status = cache.diff[lua_line]

    if status == "added" then
        return "\x1b[32m+\x1b[0m "  -- Green +
    elseif status == "modified" then
        return "\x1b[33m~\x1b[0m "  -- Yellow ~
    elseif status == "deleted" then
        return "\x1b[31m-\x1b[0m "  -- Red -
    else
        return "  "
    end
end

-- Setup function
function M.setup()
    -- Register gutter renderer
    _G._gutter_renderer = M.render_gutter

    -- Initialize git for current buffer (if buffer API is available and has filename)
    if buffer and buffer.get_filename then
        local success, filename = pcall(buffer.get_filename)
        if success and filename then
            M.init_buffer(filename)
        end
    end

    editor.message("Git plugin loaded")
end

-- Auto-setup on load
M.setup()

return M
